### PR TITLE
Story 2560. Added Custom WCS validator and related util functions.

### PR DESCRIPTION
**Pull request by yeunga**

### CODE REVIEW

File: caom2utils/caom2utils/wcsvalidator.py [hunk 1]
```diff
@@ -298,6 +308,77 @@ def _validate_polarization_wcs(polarization_wcs):
...
+        axis_list = ["" for x in range(chunk.naxis + 1)]
+        attr_dict = vars(chunk)
+        for key in attr_dict.keys():
+            if key[0] != '_' and 'axis' in key:
+                value = attr_dict.get(key)
+                if value is not None and value <= chunk.naxis:
+                    # Ignore axes greater than naxis: situation is allowed
+                    if axis_list[value] is not None and len(axis_list[value].strip()) > 0:
+                        # Flag duplicate axis definitions
+                        error_msg
```
Comment (jburke-cadc) [inline caom2utils/caom2utils/wcsvalidator.py]: Should error_msg be set to a message?

File: caom2utils/caom2utils/wcs_util.py [hunk 2]
```diff
@@ -165,6 +172,411 @@ def validate_wcs(temporal_wcs):
...
+        :param si An Interval
+        :param samples A list of Intervals
+        :param union_scale A float
+        """
+        snew = si
+        if len(samples) > 0:
+            f = union_scale * (si.upper - si.lower)
+            a = si.lower - f
+            b = si.upper + f
+            tmp = []
```
[the remark this answers is not in the document]
Comment (jburke-cadc) [reply caom2utils/caom2utils/wcs_util.py]: Ignore the above. Saw an example that explains scaling is the percentage to apply extending the interval. Makes sense now.

File: caom2/caom2/chunk.py [hunk 3]
```diff
@@ -988,7 +988,8 @@ def __init__(self, axis):
         """Set up a CoordAxis1D object to represent the Custom Axis.
 
         """
-        self.axis = axis
+        caom_util.type_check(axis, wcs.CoordAxis1D, 'axis', override=False)
+        self._axis = axis
```
[the remark this answers is not in the document]
Comment (yeunga) [reply caom2/caom2/chunk.py]: Yes. This is not the usual pattern because axis is not mutable.

[the remark this answers is not in the document]
Comment (yeunga) [reply caom2/caom2/chunk.py]: There is no setter for axis.

File: caom2utils/caom2utils/wcs_util.py [hunk 4]
```diff
@@ -165,6 +172,411 @@ def validate_wcs(temporal_wcs):
             raise ValueError(sb)
 
 
+class CustomUtil:
+    ctype_cunit_map = {"FARADAY": "rad/m**2", "RM": "rad/m**2"}
+
+    def __init__(self):
+        pass
+
```
[the remark this answers is not in the document]
Comment (yeunga) [reply caom2utils/caom2utils/wcs_util.py]: Yes, it is empty. It is probably not necessary; however I am just following the same pattern in the existing classes in wcs_util.py.

File: caom2utils/caom2utils/wcs_util.py [hunk 5]
```diff
@@ -165,6 +172,411 @@ def validate_wcs(temporal_wcs):
...
+        snew = si
+        if len(samples) > 0:
+            f = union_scale * (si.upper - si.lower)
+            a = si.lower - f
+            b = si.upper + f
+            tmp = []
+            # Find intervals that overlap the new one, move from samples -> tmp
+            for sample in samples:
+                f = union_scale * sample.upper - sample.lower
+                c = sample.lower - f
```
[the remark this answers is not in the document]
Comment (yeunga) [reply caom2utils/caom2utils/wcs_util.py]: Good catch. Thank you!

File: caom2utils/caom2utils/wcs_util.py [hunk 6]
```diff
@@ -165,6 +172,411 @@ def validate_wcs(temporal_wcs):
...
+            return False
+
+        if atype == matches:
+            logger.debug("use_chunk=True: Artifact.product_type={}"\
+                         .format(atype))
+            return True
+
+        logger.debug("use_chunk=False: product_type={},{},{}"\
+                     .format(atype, ptype, ctype))
+        return False
```
[the remark this answers is not in the document]
Comment (yeunga) [reply caom2utils/caom2utils/wcs_util.py]: I think they should all be the same.

File: caom2utils/caom2utils/wcs_util.py [hunk 7]
```diff
@@ -165,6 +172,411 @@ def validate_wcs(temporal_wcs):
...
+        """ val2pix calculates pixel from value
+
+        :param wcs A CustomWCS
+        :param func A CoordFunction1D
+        :param val A float
+        :return A float
+        """
+        CustomUtil.validate_wcs(wcs)
+        ref_val = func.ref_coord.val
+        return func.ref_coord.pix + (val - ref_val) / func.delta
```
[the remark this answers is not in the document]
Comment (yeunga) [reply caom2utils/caom2utils/wcs_util.py]: Yes, it is supposed to be the 2nd answer.

Checked with the Java version just now. 
---
    static double val2pix(CoordFunction1D function, double val) {
        double refVal = function.getRefCoord().val;
        return function.getRefCoord().pix + (val - refVal) / function.getDelta();
    }
---

File: caom2utils/caom2utils/wcs_util.py [hunk 8]
```diff
@@ -165,6 +172,411 @@ def validate_wcs(temporal_wcs):
...
+        :return An Interval
+        """
+        union_scale = float(0.02)
+        subs = []
+        for a in artifacts:
+            for p_key in a.parts:
+                p = a.parts[p_key]
+                for c in p.chunks:
+                    if CustomUtil._use_chunk(a.product_type, p.product_type, c.product_type, product_type):
+                        current_ctype = c.custom.axis.axis.ctype
```
[the remark this answers is not in the document]
Comment (yeunga) [reply caom2utils/caom2utils/wcs_util.py]: Yes, it is possible. Added check for c and c.custom to be None. Thanks.

File: caom2utils/caom2utils/wcs_util.py [hunk 9]
```diff
@@ -165,6 +172,411 @@ def validate_wcs(temporal_wcs):
...
+        sw = None
+        scale = float(0.0)
+        num = 0
+        for a in artifacts:
+            for p_key in a.parts:
+                p = a.parts[p_key]
+                for c in p.chunks:
+                    if CustomUtil._use_chunk(a.product_type, p.product_type, \
+                                            c.product_type, product_type):
+                        current_ctype = c.custom.axis.axis.ctype
```
[the remark this answers is not in the document]
Comment (yeunga) [reply caom2utils/caom2utils/wcs_util.py]: ditto above

File: caom2utils/caom2utils/wcsvalidator.py (hunk 1 above)
Comment (yeunga) [reply caom2utils/caom2utils/wcsvalidator.py]: Yes. Good catch!

File: caom2repo/caom2repo/core.py [hunk 10]
```diff
@@ -95,7 +95,9 @@
 
 BATCH_SIZE = int(10000)
 
-CAOM2REPO_OBS_CAPABILITY_ID =\
+CURRENT_CAOM2REPO_OBS_CAPABILITY_ID = \
+    'vos://cadc.nrc.ca~vospace/CADC/std/CAOM2Repository#obs-1.2'
+PREVIOUS_CAOM2REPO_OBS_CAPABILITY_ID =\
     'vos://cadc.nrc.ca~vospace/CADC/std/CAOM2Repository#obs-1.1'
```
Comment (andamian) [inline caom2repo/caom2repo/core.py]: These should be the keys to a dictionary. The values would be CAOM2 namespaces from `caom2.obs_reader_writer` ex for `{'...obs-1.1': caom2.obs_reader_writer.CAOM23_NAMESPACE, `..obs-1.2: caom2.obs_reader_writer.CAOM24_NAMESPACE`}
The values of the dictionary will be used when instantiating the ObservationWriter in order to send the correct version of the document to the repo.

File: caom2utils/caom2utils/wcs_util.py [hunk 11]
```diff
@@ -75,9 +75,16 @@
...
-from caom2 import shape
+from caom2 import shape, chunk, plane
+from builtins import int
+import logging
+import sys
 
-__all__ = ['TimeUtil', 'EnergyUtil', 'ORIGIN']
+
+APP_NAME = 'wcs_util'
+logger = logging.getLogger(APP_NAME)
```
[the remark this answers is not in the document]
Comment (yeunga) [reply caom2utils/caom2utils/wcs_util.py]: Changed.

File: caom2repo/setup.cfg [hunk 12]
```diff
@@ -35,7 +35,7 @@ edit_on_github = False
 github_project = opencadc/caom2tools
 install_requires = cadcutils>=1.1.11 caom2>=2.3.6
```
[the remark this answers is not in the document]
Comment (yeunga) [reply caom2repo/setup.cfg]: Changed.

File: caom2repo/setup.cfg [hunk 13]
```diff
@@ -35,7 +35,7 @@ edit_on_github = False
...
 # version should be PEP386 compatible (http://www.python.org/dev/peps/pep-0386)
-version = 1.4.3
+version = 1.4.4
```
[the remark this answers is not in the document]
Comment (yeunga) [reply caom2repo/setup.cfg]: Changed.

File: caom2repo/caom2repo/core.py (hunk 10 above)
Comment (yeunga) [reply caom2repo/caom2repo/core.py]: Changed to something similar.

File: caom2utils/caom2utils/wcs_util.py [hunk 14]
```diff
@@ -165,6 +172,448 @@ def validate_wcs(temporal_wcs):
             raise ValueError(sb)
 
 
+class CustomUtil:
```
[the remark this answers is not in the document]
Comment (yeunga) [reply caom2utils/caom2utils/wcs_util.py]: Will update once the unit tests have been fixed and the integration tests have been rerun.

[the remark this answers is not in the document]
Comment (yeunga) [reply caom2utils/caom2utils/wcs_util.py]: Renamed CustomUtil to CustomAxisUtil and added a short description.

**Pull request merged**